File: src/lexer/token.ts
import { Location } from "../shared";

export const KEYWORDS = new Set([
  "%macro",
  "%end",
  "%del",
  "%include",
  "proc",
  "unsafe",
  "inline",
  "::",
  "->",
  "const",
  "memory",
  "var",
  "override",
  "extern",
  "as",
  "assert",
  "if",
  "elif",
  "else",
  "while",
  "let",
  "cast",
  "return",
  "in",
  "do",
  "end",
]);

export enum Tokens {
  // Literals
  Int = "<int>",
  Str = "<str>",
  CStr = "<cstr>",
  Word = "<word>",
  AsmBlock = "<asm>",
  // Preprocessor Directives
  Macro = "%macro",
  EndPre = "%end",
  Del = "%del",
  Include = "%include",
  // Keywords
  Proc = "proc",
  Unsafe = "unsafe",
  Inline = "inline",
  SigIns = "::",
  SigOuts = "->",
  Const = "const",
  Memory = "memory",
  Var = "var",
  Override = "override",
  Extern = "extern",
  As = "as",
  Assert = "assert",
  If = "if",
  ElseIf = "elif",
  Else = "else",
  While = "while",
  Let = "let",
  Cast = "cast",
  Return = "return",
  In = "in",
  Do = "do",
  End = "end",
  // Special tokens
  EOF = "EOF"
}

export interface Token {
  kind: Tokens;
  loc: Location;
  value?: any;
}

File: src/lexer/token.rs
use crate::shared::location::Span;

#[derive(Clone, PartialEq, Eq, Debug)]
pub enum Tokens {
    Str(String),
    Int(i64),
    Bool(bool),
    Word(String),

    Proc,
    Do,
    End,
}

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct Token {
    pub kind: Tokens,
    pub span: Span,
}

impl Token {
    pub fn new(kind: Tokens, span: Span) -> Self {
        Self { kind, span }
    }

    pub fn to_str(&self) -> &str {
        match self.kind {
            Tokens::Str(_) => "<str>",
            Tokens::Int(_) => "<int>",
            Tokens::Bool(_) => "<bool>",
            Tokens::Word(_) => "<word>",
            Tokens::Proc => "proc",
            Tokens::Do => "do",
            Tokens::End => "end"
        }
    }
}